File: src-python/hf/basis_stog.py
import math
from functools import cached_property
from pydantic import BaseModel, ConfigDict, Field
from functools import cached_property
from typing import Callable, Literal
# from source.hfp import GTO_LIB
import numpy as np
from scipy.special import factorial2
from functools import partial
from scipy.special import hyp1f1


class STOGPrimitive(BaseModel):
    """Gaussian Type Orbital (GTO) basis function."""
    coords: tuple[float, float, float] = (0.0, 0.0, 0.0)  # (X, Y, Z)

    # Exponent and angular momentum quantum numbers for the GTO (s, p, d, etc.)
    alpha: float
    nx: int = 0
    ny: int = 0
    nz: int = 0

    @property
    def l(self) -> int:
        """Calculate the total angular momentum quantum number."""
        return self.nx + self.ny + self.nz

    @cached_property
    def N(self) -> float:
        """Normalization constant for the cartisian GTO. https://en.wikipedia.org/wiki/Gaussian_orbital#Cartesian_coordinates"""
        prefactor = (2 * self.alpha / math.pi) ** (3 / 4)
        numerator = (8 * self.alpha) ** (self.nx + self.ny + self.nz) * math.factorial(self.nx) * math.factorial(self.ny) * math.factorial(self.nz)
        denominator = math.factorial(2*self.nx) * math.factorial(2*self.ny) * math.factorial(2*self.nz)
        return prefactor * math.sqrt(numerator/ denominator)
    

    def Y(self, x: float, y: float , z: float) -> float:
        """Calculate the angular part of the GTO in cartesian coordinates."""
        return (x - self.coords[0]) ** self.nx * (y - self.coords[1]) ** self.ny * (z - self.coords[2]) ** self.nz

    def r2(self, x: float, y: float, z: float) -> float:
        """Calculate the squared distance from the center of the GTO."""
        dist_x = x - self.coords[0]
        dist_y = y - self.coords[1]
        dist_z = z - self.coords[2]
        return dist_x ** 2 + dist_y ** 2 + dist_z ** 2

    def r(self, x: float, y: float, z: float) -> float:
        """Calculate the distance from the center of the GTO."""
        return math.sqrt(self.r2(x, y, z))

    def __call__(self, x: float, y: float, z: float) -> float:
        """Evaluate the GTO at a given point (x, y, z)."""
        R = math.exp(-self.alpha * self.r2(x, y, z))
        return self.N * R * self.Y(x, y, z) # Use cartesian coordinates

class STOGOrbital(BaseModel):
    """STO-nG for an atomic orbital (1s, 2s, 2p, etc.), which is a linear combination of n GTOs."""

    model_config = ConfigDict(arbitrary_types_allowed=True)  # Forbid extra fields not defined in the model
    
    type: str
    atom: str | None = None
    coords: tuple[float, float, float] = (0.0, 0.0, 0.0)

    cc: list[float] = Field(..., description="Contraction coefficients for the GTOs in the STO-nG expansion")
    alpha: list[float] = Field(..., description="Exponent coefficients for the GTOs in the STO-nG expansion")
    nx: list[int] = Field(..., description="Angular momentum quantum numbers in x direction for the GTOs in the STO-nG expansion")
    ny: list[int] = Field(..., description="Angular momentum quantum numbers in y direction for the GTOs in the STO-nG expansion")
    nz: list[int] = Field(..., description="Angular momentum quantum numbers in z direction for the GTOs in the STO-nG expansion")

    @property
    def l(self) -> int:
        """Calculate the total angular momentum quantum number."""
        return self.nx[0] + self.ny[0] + self.nz[0]

    @cached_property
    def gtos(self) -> list[STOGPrimitive]:
        """Calculate the GTOs that make up the STO-nG orbital."""
        return [STOGPrimitive(coords=self.coords, alpha=self.alpha[i], nx=self.nx[i], ny=self.ny[i], nz=self.nz[i]) for i in range(len(self.cc))]
    
    @cached_property
    def N(self) -> float:
        """The normalization constant for the STO-nG orbital."""
        return [gto.N for gto in self.gtos]
    
    def __call__(self, x: float, y: float, z: float) -> float:
        """Evaluate the STO-nG orbital at a given point (x, y, z)."""
        return sum(self.cc[i] * self.gtos[i](x, y, z) for i in range(len(self.cc))) # Use cartesian coordinates

class STOGIntegrator:
    """
    Integrator class for calculating integrals between STO-nG orbitals.
    The idea is for each S(i,j) where i is the ith atomic orbital, we compute
    a mxn matrix where each element is an integral between Gausian primitive
    m of atomic orbital i and Gaussian primitive n of atomic orbital j.
    We than sum over all the elements in the matrix to get S(i,j).
    We do this for all pairs of atomic orbitals to get the full overlap matrix S.
    """

    def Sij(self, orb1: STOGOrbital, orb2: STOGOrbital) -> float:
        """Calculate the overlap integral between two STO-nG orbitals."""

        def compute_Sab(gto1: STOGPrimitive, gto2: STOGPrimitive) -> float:
            """Calculate the overlap integral S_ab between two GTO primitives. (eq.101)"""
            prefactor = (math.pi / (gto1.alpha + gto2.alpha)) ** (3 / 2)
            s_x = self.compute_E(gto1.coords[0], gto2.coords[0], gto1.alpha, gto2.alpha, gto1.nx, gto2.nx, 0)
            s_y = self.compute_E(gto1.coords[1], gto2.coords[1], gto1.alpha, gto2.alpha, gto1.ny, gto2.ny, 0)
            s_z = self.compute_E(gto1.coords[2], gto2.coords[2], gto1.alpha, gto2.alpha, gto1.nz, gto2.nz, 0)
            return prefactor * s_x * s_y * s_z

        sij = 0.0
        for u, cc1 in enumerate(orb1.cc):
            for v, cc2 in enumerate(orb2.cc):
                sij += cc1 * cc2 * orb1.gtos[u].N * orb2.gtos[v].N * compute_Sab(orb1.gtos[u], orb2.gtos[v])
        return 0.0 if math.isclose(sij, 0.0) else sij
    
    def Tij(self, orb1: STOGOrbital, orb2: STOGOrbital) -> float:
        """Calculate the kinetic energy integral between two STO-nG orbitals."""

        def compute_D(A: float, B: float, alpha: float, beta: float, ai: int, bi: int, t: int) -> float:
            """ Returns the Dawson function D_n(T) """
            p = alpha + beta
            if t == 0: # eq. 113 
                return self.compute_E(A, B, alpha, beta, ai, bi, 0) * np.sqrt(np.pi/(p))
            term1 = bi * compute_D(A, B, alpha, beta, ai, bi-1, t-1)
            term2 = 2*beta*compute_D(A, B, alpha, beta, ai, bi+1, t-1)
            return term1 - term2 # eq. 114

        def compute_Tab(orb1: STOGOrbital, orb2: STOGOrbital) -> float:
            """Calculate the kinetic energy integral T_ab between two GTO primitives. (eq. 121)"""
            term1 = compute_D(orb1.coords[0], orb2.coords[0], orb1.alpha, orb2.alpha, orb1.nx, orb2.nx, 2)
            term2 = compute_D(orb1.coords[1], orb2.coords[1], orb1.alpha, orb2.alpha, orb1.ny, orb2.ny, 0)
            term3 = compute_D(orb1.coords[2], orb2.coords[2], orb1.alpha, orb2.alpha, orb1.nz, orb2.nz, 0)
            term4 = compute_D(orb1.coords[0], orb2.coords[0], orb1.alpha, orb2.alpha, orb1.nx, orb2.nx, 0)
            term5 = compute_D(orb1.coords[1], orb2.coords[1], orb1.alpha, orb2.alpha, orb1.ny, orb2.ny, 2)
            term6 = compute_D(orb1.coords[2], orb2.coords[2], orb1.alpha, orb2.alpha, orb1.nz, orb2.nz, 0)
            term7 = compute_D(orb1.coords[0], orb2.coords[0], orb1.alpha, orb2.alpha, orb1.nx, orb2.nx, 0)
            term8 = compute_D(orb1.coords[1], orb2.coords[1], orb1.alpha, orb2.alpha, orb1.ny, orb2.ny, 0)
            term9 = compute_D(orb1.coords[2], orb2.coords[2], orb1.alpha, orb2.alpha, orb1.nz, orb2.nz, 2)
            return -0.5 * (term1*term2*term3 + term4*term5*term6 + term7*term8*term9)

        tij = 0.0
        for u, cc1 in enumerate(orb1.cc):
            for v, cc2 in enumerate(orb2.cc):
                tij += cc1 * cc2 * orb1.gtos[u].N * orb2.gtos[v].N * compute_Tab(orb1.gtos[u], orb2.gtos[v])
        return 0.0 if math.isclose(tij, 0.0) else tij
    
    def VijR(self, orb1: STOGOrbital, orb2: STOGOrbital, R: tuple[float, float, float]) -> float:
        """Calculate the electron-nuclear attraction integral between two STO-nG orbitals for a single nucleus."""

        def compute_Vab(gto1: STOGPrimitive, gto2: STOGPrimitive, R: tuple[float, float, float]) -> float:
            """Calculate the nuclear attraction integral between two STO-nG orbitals for a single nucleus. (eq. 199-204)"""
            p = gto1.alpha + gto2.alpha
            P = [self.compute_Pi(gto1, gto2, component) for component in ['x', 'y', 'z']] # Gaussian composite center
            val = 0.0
            for t in range(gto1.nx+gto2.nx+1):
                for u in range(gto1.ny+gto2.ny+1):
                    for v in range(gto1.nz+gto2.nz+1):
                        val += self.compute_E(gto1.coords[0], gto2.coords[0], gto1.alpha, gto2.alpha, gto1.nx, gto2.nx, t) * \
                            self.compute_E(gto1.coords[1], gto2.coords[1], gto1.alpha, gto2.alpha, gto1.ny, gto2.ny, u) * \
                            self.compute_E(gto1.coords[2], gto2.coords[2], gto1.alpha, gto2.alpha, gto1.nz, gto2.nz, v) * \
                            self.compute_R(t, u, v, 0, p, P, R)
            return 2*np.pi/p  * val

        vijr = 0.0
        for u, cc1 in enumerate(orb1.cc):
            for v, cc2 in enumerate(orb2.cc):
                vijr += orb1.gtos[u].N*orb2.gtos[v].N*cc1*cc2*compute_Vab(orb1.gtos[u], orb2.gtos[v], R)
        return 0.0 if math.isclose(vijr, 0.0) else vijr

    def Vijkl(self, orb1: STOGOrbital, orb2: STOGOrbital, orb3: STOGOrbital, orb4: STOGOrbital) -> float:
        """Calculate the Coulomb repulsion integral between two STO-nG orbitals."""

        def compute_Vabcd(gto1: STOGPrimitive, gto2: STOGPrimitive, gto3: STOGPrimitive, gto4: STOGPrimitive) -> float:
            """Calculate the electron-electron repulsion integral between two STO-nG orbitals."""
            p = gto1.alpha + gto2.alpha
            q = gto3.alpha + gto4.alpha
            alpha = p * q / (p + q)
            P = [self.compute_Pi(gto1, gto2, component) for component in ['x', 'y', 'z']]
            Q = [self.compute_Pi(gto3, gto4, component) for component in ['x', 'y', 'z']]

            val = 0.0
            for t in range(gto1.nx+gto2.nx+1):
                for u in range(gto1.ny+gto2.ny+1):
                    for v in range(gto1.nz+gto2.nz+1):
                        for tau in range(gto3.nx+gto4.nx+1):
                            for nu in range(gto3.ny+gto4.ny+1):
                                for phi in range(gto3.nz+gto4.nz+1):
                                    term1 = self.compute_E(gto1.coords[0], gto2.coords[0], gto1.alpha, gto2.alpha, gto1.nx, gto2.nx, t)
                                    term2 = self.compute_E(gto1.coords[1], gto2.coords[1], gto1.alpha, gto2.alpha, gto1.ny, gto2.ny, u)
                                    term3 = self.compute_E(gto1.coords[2], gto2.coords[2], gto1.alpha, gto2.alpha, gto1.nz, gto2.nz, v)
                                    term4 = self.compute_E(gto3.coords[0], gto4.coords[0], gto3.alpha, gto4.alpha, gto3.nx, gto4.nx, tau)
                                    term5 = self.compute_E(gto3.coords[1], gto4.coords[1], gto3.alpha, gto4.alpha, gto3.ny, gto4.ny, nu)
                                    term6 = self.compute_E(gto3.coords[2], gto4.coords[2], gto3.alpha, gto4.alpha, gto3.nz, gto4.nz, phi)
                                    term7 = np.power(-1, tau+nu+phi)
                                    term8 = self.compute_R(t+tau, u+nu, v+phi, 0, alpha, P, Q)
                                    val += term1*term2*term3*term4*term5*term6*term7*term8
            prefactor = 2 * math.pi**(5/2) / (p * q * math.sqrt(p + q))
            return prefactor * val
        
        vijkl = 0.0
        for m, cc1 in enumerate(orb1.cc):
            for n, cc2 in enumerate(orb2.cc):
                for u, cc3 in enumerate(orb3.cc):
                    for v, cc4 in enumerate(orb4.cc):
                        norms = orb1.gtos[m].N * orb2.gtos[n].N * orb3.gtos[u].N * orb4.gtos[v].N
                        coefs = cc1 * cc2 * cc3 * cc4
                        vijkl += norms * coefs * compute_Vabcd(orb1.gtos[m], orb2.gtos[n], orb3.gtos[u], orb4.gtos[v])
        return 0.0 if math.isclose(vijkl, 0.0) else vijkl


    ####### Common subroutines for Sij, Tij, and Vij #######

    def compute_E(self, A: float, B: float, alpha: float, beta: float, ai: int, bi: int, t: int) -> float:
        """ Recursive definition of Hermite Gaussian coefficients using eq. 73-75.
            A, B: origin of Gaussian 'a' and 'b'
            alpha, beta: exponent of Gaussian 'a' and 'b'
            ai, bi: orbital angular momentum number on Gaussian 'a' and 'b'. i,j in the paper; nx, ny, nz in the code.
            t: number nodes in Hermite (depends on type of integral, e.g. always zero for overlap integrals)
        """

        func = partial(self.compute_E, A=A, B=B, alpha=alpha, beta=beta)
        p = alpha + beta
        q = alpha*beta/p
        Qx = A - B
        if (t < 0) or (t > (ai + bi)):
            return 0.0
        if ai == bi == t == 0: # eq: 73
            return np.exp(-q*Qx**2)
        if bi == 0: # eq: 74
            term1 = (1/(2*p))*func(ai=ai-1,bi=bi,t=t-1)
            term2 = (q*Qx/alpha)*func(ai=ai-1,bi=bi,t=t)
            term3 = (t+1)*func(ai=ai-1,bi=bi,t=t+1)
            return term1 - term2 + term3
        # eq: 75
        term1 = (1/(2*p))*func(ai=ai,bi=bi-1,t=t-1)
        term2 = (q*Qx/beta)*func(ai=ai,bi=bi-1,t=t)
        term3 = (t+1)*func(ai=ai,bi=bi-1,t=t+1)
        return term1 + term2 + term3

    def compute_R(self, t: int,u: int, v:int, n:int, p:float, P: tuple[float, float, float], C: tuple[float, float, float]) -> float:
        """ Returns the Coulomb auxiliary Hermite integrals (eq. 190-192)
            Arguments:
                t,u,v: order of Coulomb Hermite derivative in x,y,z
                n: order of Boys function 
                C: Nuclear center C.
        """

        def boys(n: int, T: float) -> float:
            """Boys function using confluent hypergeometric function."""
            return hyp1f1(n+0.5,n+1.5,-T)/(2.0*n+1.0) 

        RPC = math.sqrt((P[0]-C[0])**2 + (P[1]-C[1])**2 + (P[2]-C[2])**2)
        val = 0.0
        if t < 0 or u < 0 or v < 0:
            return 0.0
        if t == u == v == 0: # eq. 182
            val += (-2*p)**(n)*boys(n,p*RPC*RPC)
        elif t == u == 0:
            val +=(v-1)*self.compute_R(t,u,v-2,n+1,p,P,C)
            val += (P[2]-C[2])*self.compute_R(t,u,v-1,n+1,p,P,C)
        elif t == 0:
            val += (u-1)*self.compute_R(t,u-2,v,n+1,p,P,C)
            val += (P[1]-C[1])*self.compute_R(t,u-1,v,n+1,p,P,C)
        else:
            val += (t-1)*self.compute_R(t-2,u,v,n+1,p,P,C)
            val += (P[0]-C[0])*self.compute_R(t-1,u,v,n+1,p,P,C)
        return val
    
    def compute_Pi(self, gto1: STOGPrimitive, gto2: STOGPrimitive, component: Literal['x', 'y', 'z'] = 'x') -> float:
        """Calculate the weighted center and combined exponent for the integral in a given component."""
        alpha_sum = gto1.alpha + gto2.alpha
        if component == 'x':
            return (gto1.alpha * gto1.coords[0] + gto2.alpha * gto2.coords[0]) / alpha_sum
        elif component == 'y':
            return (gto1.alpha * gto1.coords[1] + gto2.alpha * gto2.coords[1]) / alpha_sum
        else:  # component == 'z'
            return (gto1.alpha * gto1.coords[2] + gto2.alpha * gto2.coords[2]) / alpha_sum


    

File: src-rust/src/hf/basis_stog.rs
use std::f64::consts::PI;
use xsf::hyp1f1;

pub mod stog_primitive {
    pub struct STOGPrimitive {
        pub coords: (f64, f64, f64),
        pub alpha: f64,
        pub cc: f64,
        pub nx: i32,
        pub ny: i32,
        pub nz: i32,
        pub l: i32,
        pub N: f64,
    }

    impl STOGPrimitive {
        pub fn new(
            coords: (f64, f64, f64),
            alpha: f64,
            cc: f64,
            nx: i32,
            ny: i32,
            nz: i32,
        ) -> Self {

            fn factorial(num: u32) -> f64 {
                (1..=num).map(|x| x as f64).product()
            }
            let prefactor = (2.0 * alpha / PI).powf(0.75);
            let numerator = (8.0 * alpha).powi((nx + ny + nz) as i32) * factorial(nx as u32) * factorial(ny as u32) * factorial(nz as u32);
            let denominator = factorial(2 * nx as u32) * factorial(2 * ny as u32) * factorial(2 * nz as u32);
            let N = prefactor * (numerator / denominator).sqrt();
            StoGPrimitive {
                coords,
                alpha,
                cc,
                nx,
                ny,
                nz,
                nx + ny + nz,
                N,
            }
        }
    }
}

pub mod stog_orbital {
    use super::stog_primitive::STOGPrimitive;

    pub struct STOGOrbital {
        pub atom: String,
        pub shell: String,
        pub coords: (f64, f64, f64),
        pub gtos: Vec<STOGPrimitive>,
    }

    impl STOGOrbital {
        pub fn new(atom: String, shell: String, coords: (f64, f64, f64), cc: Vec<f64>, alpha: Vec<f64>, nx: Vec<u32>, ny: Vec<u32>, nz: Vec<u32>) -> Self {
            let mut gtos = Vec::new();
            for i in 0..cc.len() {
                gtos.push(STOGPrimitive::new(coords, alpha[i], cc[i], nx[i], ny[i], nz[i]));
            }
            STOGOrbital { atom, shell, coords, gtos }
        }
    }
}

pub mod stog_integrator {
    use super::stog_primitive::STOGPrimitive;
    use super::stog_orbital::STGOrbital;
    
    pub fn compute_Sij(orb1: &STOGOrbital, orb2: &STOGOrbital) -> f64 {

        fn compute_Sab(gto1: &STOGPrimitive, gto2: &STOGPrimitive) -> f64 {
            // Overlap integral between two STO-G primitives
            let prefactor = (PI / (gto1.alpha + gto2.alpha)).powf(1.5);
            let s_x = compute_E(gto1.coords.0, gto2.coords.0, gto1.alpha, gto2.alpha, gto1.nx as i32, gto2.nx as i32, 0);
            let s_y = compute_E(gto1.coords.1, gto2.coords.1, gto1.alpha, gto2.alpha, gto1.ny as i32, gto2.ny as i32, 0);
            let s_z = compute_E(gto1.coords.2, gto2.coords.2, gto1.alpha, gto2.alpha, gto1.nz as i32, gto2.nz as i32, 0);
            return prefactor * s_x * s_y * s_z
        }

        let mut sij = 0.0;
        for u in 0..orb1.gtos.len() {
            for v in 0..orb2.gtos.len() {
                let gto1 = &orb1.gtos[u];
                let gto2 = &orb2.gtos[v];
                sij += gto1.N * gto2.N * gto1.cc * gto2.cc * compute_Sab(gto1, gto2);
            }
        }
        return sij;
    }

    pub fn compute_Tij(orb1: &STOGOrbital, orb2: &STOGOrbital) -> f64 {
        let tij = 0.0;

        fn compute_D(A: f64, B: f64, alpha: f64, beta: f64, ai: i32, bi: i32, t: i32) -> f64 {
            let p = alpha + beta;
            if t == 0 {return compute_E(A, B, alpha, beta, ai, bi, 0) * (PI/p).powf(0.5);}
            let term1 = bi as f64 * compute_D(A, B, alpha, beta, ai, bi-1, t-1);
            let term2 = 2.0 * beta * compute_D(A, B, alpha, beta, ai, bi+1, t-1);
            return term1 - term2;
        }

        fn compue_Tab(orb1: &STOGPrimitive, orb2: &STOGPrimitive) -> f64 {
            let term1 = compute_D(orb1.coords.0, orb2.coords.0, orb1.alpha, orb2.alpha, orb1.nx, orb2.nx, 2);
            let term2 = compute_D(orb1.coords.1, orb2.coords.1, orb1.alpha, orb2.alpha, orb1.ny, orb2.ny, 2);
            let term3 = compute_D(orb1.coords.2, orb2.coords.2, orb1.alpha, orb2.alpha, orb1.nz, orb2.nz, 2);
            let term4 = compute_D(orb1.coords.0, orb2.coords.0, orb1.alpha, orb2.alpha, orb1.nx, orb2.nx, 0);
            let term5 = compute_D(orb1.coords.1, orb2.coords.1, orb1.alpha, orb2.alpha, orb1.ny, orb2.ny, 0);
            let term6 = compute_D(orb1.coords.2, orb2.coords.2, orb1.alpha, orb2.alpha, orb1.nz, orb2.nz, 0);
            return -0.5 * (term1*term5*term6 + term4*term2*term6 + term4*term5*term3);
        }
        
        for u in 0..orb1.gtos.len() {
            for v in 0..orb2.gtos.len() {
                let gto1 = &orb1.gtos[u];
                let gto2 = &orb2.gtos[v];
                tij += gto1.N * gto2.N * gto1.cc * gto2.cc * compue_Tab(gto1, gto2);
            }
        }
        return tij;
    }

    pub fn compute_VijR(orb1: &STOGOrbital, orb2: &STOGOrbital, R: (f64, f64, f64)) -> f64 {
        // Placeholder for nuclear attraction integral computation between STO-G orbitals
        fn compute_Vab(gto1: &STOGPrimitive, gto2: &STOGPrimitive, R: (f64, f64, f64)) -> f64 {
            let p = gto1.alpha + gto2.alpha;
            let P = (
                (gto1.alpha * gto1.coords.0 + gto2.alpha * gto2.coords.0) / p,
                (gto1.alpha * gto1.coords.1 + gto2.alpha * gto2.coords.1) / p,
                (gto1.alpha * gto1.coords.2 + gto2.alpha * gto2.coords.2) / p,
            );

            let mut val = 0.0;
            for t in 0..=(gto1.nx + gto2.nx) {
                for u in 0..=(gto1.ny + gto2.ny) {
                    for v in 0..=(gto1.nz + gto2.nz) {
                        let E_x = compute_E(gto1.coords.0, gto2.coords.0, gto1.alpha, gto2.alpha, gto1.nx, gto2.nx, t);
                        let E_y = compute_E(gto1.coords.1, gto2.coords.1, gto1.alpha, gto2.alpha, gto1.ny, gto2.ny, u);
                        let E_z = compute_E(gto1.coords.2, gto2.coords.2, gto1.alpha, gto2.alpha, gto1.nz, gto2.nz, v);
                        let R_tuv = compute_R(t as i32, u as i32, v as i32, 0, p, P, R);
                        val += E_x * E_y * E_z * R_tuv;
                    }
                }
            }
            return (2.0 * PI / p) * val;
        }

        let mut vijr = 0.0;
        for u in 0..orb1.gtos.len() {
            for v in 0..orb2.gtos.len() {
                let gto1 = &orb1.gtos[u];
                let gto2 = &orb2.gtos[v];
                vijr += gto1.N * gto2.N * gto1.cc * gto2.cc * compute_Vab(gto1, gto2, R);
            }
        }
        return vijr;
    }

    pub fn compute_Vijkl(orb1: &STOGOrbital, orb2: &STOGOrbital, orb3: &STOGOrbital, orb4: &STOGOrbital) -> f64 {
        fn compute_Vabcd(gto1: &STOGPrimitive, gto2: &STOGPrimitive, gto3: &STOGPrimitive, gto4: &STOGPrimitive) -> f64 {
            p = gto1.alpha + gto2.alpha;
            q = gto3.alpha + gto4.alpha;
            P = (
                (gto1.alpha * gto1.coords.0 + gto2.alpha * gto2.coords.0) / p,
                (gto1.alpha * gto1.coords.1 + gto2.alpha * gto2.coords.1) / p,
                (gto1.alpha * gto1.coords.2 + gto2.alpha * gto2.coords.2) / p,
            );
            Q = (
                (gto3.alpha * gto3.coords.0 + gto4.alpha * gto4.coords.0) / q,
                (gto3.alpha * gto3.coords.1 + gto4.alpha * gto4.coords.1) / q,
                (gto3.alpha * gto3.coords.2 + gto4.alpha * gto4.coords.2) / q,
            );

            let mut val = 0.0;
            for t in 0..=(gto1.nx + gto2.nx) {
                for u in 0..=(gto1.ny + gto2.ny) {
                    for v in 0..=(gto1.nz + gto2.nz) {
                        for tau in 0..=(gto3.nx + gto4.nx) {
                            for nu in 0..=(gto3.ny + gto4.ny) {
                                for phi in 0..=(gto3.nz + gto4.nz) {
                                    let term1 = compute_E(gto1.coords.0, gto2.coords.0, gto1.alpha, gto2.alpha, gto1.nx, gto2.nx, t);
                                    let term2 = compute_E(gto1.coords.1, gto2.coords.1, gto1.alpha, gto2.alpha, gto1.ny, gto2.ny, u);
                                    let term3 = compute_E(gto1.coords.2, gto2.coords.2, gto1.alpha, gto2.alpha, gto1.nz, gto2.nz, v);
                                    let term4 = compute_E(gto3.coords.0, gto4.coords.0, gto3.alpha, gto4.alpha, gto3.nx, gto4.nx, tau);
                                    let term5 = compute_E(gto3.coords.1, gto4.coords.1, gto3.alpha, gto4.alpha, gto3.ny, gto4.ny, nu);
                                    let term6 = compute_E(gto3.coords.2, gto4.coords.2, gto3.alpha, gto4.alpha, gto3.nz, gto4.nz, phi);
                                    let term7 = (-1.0 as f64).powi((tau + nu + phi) as i32);
                                    let term8 = compute_R((t+tau) as i32, (u + nu) as i32, (v + phi) as i32, 0, p * q / (p + q), P, Q);
                                    val += term1 * term2 * term3 * term4 * term5 * term6 * term7 * term8;
                                }
                            }
                        }
                    }
                }
            }
            return (2.0 * PI.powf(2.5) / (p * q * (p + q).sqrt())) * val;
        }

        let mut vijkl = 0.0;
        for a in 0..orb1.gtos.len() {
            for b in 0..orb2.gtos.len() {
                for c in 0..orb3.gtos.len() {
                    for d in 0..orb4.gtos.len() {
                        let gto1 = &orb1.gtos[a];
                        let gto2 = &orb2.gtos[b];
                        let gto3 = &orb3.gtos[c];
                        let gto4 = &orb4.gtos[d];
                        vijkl += gto1.N * gto2.N * gto3.N * gto4.N * gto1.cc * gto2.cc * gto3.cc * gto4.cc * compute_Vabcd(gto1, gto2, gto3, gto4);
                    }
                }
            }
        }
        return vijkl;
    }

    fn compute_E(A: f64, B: f64, alpha: f64, beta: f64, ai: i32, bi: i32, t: i32) -> f64 {
        let p = alpha + beta;
        let q = alpha * beta / p;
        let Qx = A - B;
        if t < 0 || t > ai + bi {return 0.0;}
        if ai == 0 && bi == 0 && t == 0 {return (-q * Qx * Qx).exp();}
        if bi == 0 {
            let term1 = (1.0/(2.0*p)) * compute_E(A, B, alpha, beta, ai-1, 0, t-1);
            let term2 = - (q * Qx / alpha) * compute_E(A, B, alpha, beta, ai-1, 0, t);
            let term3 = (t + 1) * compute_E(A, B, alpha, beta, ai-1, 0, t+1);
            return term1 + term2 + term3;   
        }
        let term1 = (1.0/(2.0*p)) * compute_E(A, B, alpha, beta, ai, bi-1, t-1);
        let term2 = (q * Qx / beta) * compute_E(A, B, alpha, beta, ai, bi-1, t);
        let term3 = (t + 1) * compute_E(A, B, alpha, beta, ai, bi-1, t+1);
        return term1 + term2 + term3;   
    }

    fn compute_R(t: i32, u: i32, v: i32, n: i32, p: f64, P: (f64, f64, f64), C: (f64, f64, f64)) -> f64 {

        fn boys(n: u32, T: f64) -> f64 {
            return hyp1f1(n as f64 + 0.5, n as f64 + 1.5, -T) / (2.0 * n as f64 + 1.0);
        }
        
        let RPC = ((P.0 - C.0).powi(2) + (P.1 - C.1).powi(2) + (P.2 - C.2).powi(2)).sqrt();
        let mut val = 0.0;

        if t < 0 || u < 0 || v < 0 {
            return 0.0;
        }
        if t == 0 && u == 0 && v == 0{
            val += (-2.0 * p).powi(n + 1) * boys(n,p * RPC * RPC);
        }
        else if t == 0 && u == 0 {
            val += (v-1) as f64 * compute_R(0, 0, v-2, n+1, p, P, C);
            val += (P.2 - C.2) * compute_R(0, 0, v-1, n+1, p, P, C);
        }
        else if t == 0 {
            val += (u-1) as f64 * compute_R(0, u-2, v, n+1, p, P, C);
            val += (P.1 - C.1) * compute_R(0, u-1, v, n+1, p, P, C);
        }
        else {
            val += (t-1) as f64 * compute_R(t-2, u, v, n+1, p, P, C);
            val += (P.0 - C.0) * compute_R(t-1, u, v, n+1, p, P, C);
        }
        return val;
    }

}